### PR TITLE
Async mode for dnn's object detection sample

### PR DESCRIPTION
**Pull request by dkurt**
### This pullrequest changes

resolves https://github.com/opencv/opencv/issues/13667

Camera with 60fps @ 1280x720 is used (Logitech C922 PRO)

| Target | `--async=0` | `--async=1` | `--async=2` | `--async=3` | `--async=4` |
|------|-----|-----|-----|----|---|
| NCS, USB 3.0 | 11.6 | 9.9 | 11.9 | 11.9 | 11.9 |
| NCS2, USB 3.0 | 26.43 | 19.9 | 39.5 | **53.2** | 48.6 |

CPU: Intel® Core™ i7-6700K CPU @ 4.00GHz × 8
NCS: Intel® Movidius™ Neural Compute Stick
NCS2: Intel® Neural Compute Stick 2

* All the numbers are frames per second which are processed by network (FPS):

```
$ python object_detection.py opencv_fd --target 3
```

```
allow_multiple_commits=1
buildworker:Custom=linux-2
docker_image:Custom=ubuntu-openvino:16.04
build_parallel_tests:Custom=1
test_modules:Custom=dnn
test_opencl:Custom=ON
test_bigdata:Custom=1
test_filter:Custom=*
```

### CODE REVIEW

File: samples/dnn/object_detection.py [hunk 1]
```diff
@@ -1,6 +1,13 @@
 import cv2 as cv
 import argparse
 import numpy as np
+import sys
+import time
+from threading import Thread
+if sys.version[0] == '2':
+    import Queue as queue
+else:
+    import queue as queue
```
Comment (alalek) [inline samples/dnn/object_detection.py]: Just `import queue`

File: samples/dnn/object_detection.py [hunk 2]
```diff
@@ -173,32 +183,122 @@ def callback(pos):
...
+                net.setInput(np.array([[inpHeight, inpWidth, 1.6]], dtype=np.float32), 'im_info')
+
+            if args.async:
+                futureOutputs.append(net.forwardAsync())
+            else:
+                outs = net.forward(outNames)
+                predictionsQueue.put(np.copy(outs))
+
+        while futureOutputs and futureOutputs[0].wait_for(0) == 0:
+            out = futureOutputs[0].get()
```
Comment (alalek) [inline samples/dnn/object_detection.py]: BTW, callback-based approach would look better here (instead of pulling).

Comment (dkurt) [reply samples/dnn/object_detection.py]: For now it adds outputs in the order of frames. With callbacks we will need to introduce something like frames indices.

**Comment by dkurt** [comment]
Ported https://github.com/opencv/opencv/pull/14538 here.

**Comment by alalek** [comment]
Please fix this warning:

> object_detection.cpp(85): warning C4244: 'initializing': conversion from 'double' to 'float'

**Comment by AlexeyAB** [comment]
@dkurt 
This is a very good improvement for detection performance.

* It works well for VideoCamera, because FPS is limited. 

* But it works a bit strange for video files - very high capture_fps = 256 FPS, a lot of skipped frames. A five-minute video file is played in a several seconds. Everything flickers very fast on the screen.
Perhaps better to limit the capture_fps = `max(detection_fps, videofile_fps)`.

skip

**Comment by dkurt** [comment]
@AlexeyAB, thanks! That's a good point. Feel free to propose a PR or open an issue.

**Comment by lblq2008** [comment]
lblq2011@163.com
 
From: Dmitry Kurtaev
Date: 2019-07-15 11:10
To: opencv/opencv
CC: Subscribed
Subject: Re: [opencv/opencv] Async mode for dnn's object detection sample (#14516)
@AlexeyAB, thanks! That's a good point. Feel free to propose a PR or open an issue.

**Pull request merged**